File: python/CardGameEx/Ex06b.py
# Ex06b.py

from ch.aplu.jgamegrid import Location
from ch.aplu.jcardgame import Deck, Hand, CardGame, StackLayout, TargetArea, CardListener
from ch.aplu.jcardgame.Hand import SortType
from ch.aplu.util import Monitor
from enum import enum

# ------------------ class MyCardValues ------------------------------------
class MyCardValues(Deck.CardValues):
   def values(self, Suit):
      return [11, 4, 3, 2, 10]

# ------------------ class BitsCardListener --------------------------------
class BitsCardListener(CardListener):
   def atTarget(self, card, loc):
      global targetCount
      targetCount += 1
      if targetCount == nbPlayers:
         Monitor.wakeUp()  # All cards in stock->continue

# ------------------ class HandsCardListener -------------------------------
class HandsCardListener(CardListener):
   def __init__(self, i):
      self.i = i

   def leftDoubleClicked(self, card):
      global x
      global currentPlayer
      global nbMovesInRound
      card.setVerso(False)
      card.transfer(bids[self.i], True)
      hands[currentPlayer].setTouchEnabled(False)
      currentPlayer = (currentPlayer + 1) % nbPlayers
      
      if nbMovesInRound == 3:
         cg.setStatusText("Evaluating round...")
         nbMovesInRound = 0
         currentPlayer = transferToWinner()
         stocks[currentPlayer].draw()
      else:
         nbMovesInRound += 1

      if hands[currentPlayer].isEmpty():
         showResult()
      else:
         cg.setStatusText("Current player: " + str(currentPlayer))
         hands[currentPlayer].setTouchEnabled(True)

# ------------------ global functions --------------------------------------
def dealingOut():
   global hands
   global talon
   cg.setStatusText("Dealing out...")
   hands = deck.dealingOut(nbPlayers, 0)  # All cards in talon
   talon = hands[nbPlayers]
   talon.setVerso(True)

   for i in range(nbPlayers):
      hands[i].setView(cg, StackLayout(handLocations[i], 90 * i))
      hands[i].draw()

   talon.setView(cg, StackLayout(handLocations[nbPlayers]))
   talon.draw()

   while not talon.isEmpty():
      for i in range(nbPlayers):
         top = talon.getLast()
         if i == 0:
            top.setVerso(False)
         else:
            top.setVerso(True)
         talon.transfer(top, hands[i], True)

def initBids():
   for i in range(nbPlayers):
      bids[i] = Hand(deck)
      bids[i].setView(cg, StackLayout(bidLocations[i]))
      bids[i].addCardListener(BitsCardListener())
      hands[i].setTargetArea(TargetArea(bidLocations[i]))
      hands[i].addCardListener(HandsCardListener(i))

def initStocks():
   for i in range(nbPlayers):
      stocks[i] = Hand(deck)
      if i == 0 or i == 2:
        rotationAngle = 0
      else:
        rotationAngle = 90
      stocks[i].setView(cg, StackLayout(stockLocations[i], rotationAngle))

def transferToStock(player):
   global targetCount
   targetCount = 0
   for i in range(nbPlayers):
      bids[i].setTargetArea(TargetArea(stockLocations[player]))
      card = bids[i].getLast()
      if card == None:
         continue
      card.setVerso(True)
      bids[i].transferNonBlocking(card, stocks[player], True)
   Monitor.putSleep()
   stocks[player].draw()

def transferToWinner():
   cg.delay(3000)
   eval = Hand(deck)
   for i in range(nbPlayers):
      eval.insert(bids[i].getFirst(), False)
   nbWinner = eval.getMaxPosition(SortType.RANKPRIORITY)
   cg.setStatusText("Round winner: " + str(nbWinner))
   transferToStock(nbWinner)
   return nbWinner

def showResult():
    text = "Game over. Summary: "
    for i in range(nbPlayers):
      text += "Player # " + str(i) + ": " + str(stocks[i].getScore())
      if i < nbPlayers - 1:
        text += "; ";
    cg.setStatusText(text)

   
# ------------------------ main --------------------------------------------
Suit = enum("SPADES", "HEARTS", "DIAMONDS", "CLUBS")
Rank = enum("ACE", "KING", "QUEEN", "JACK", "TEN")
deck = Deck(Suit.values(), Rank.values(), "cover", MyCardValues())

nbPlayers = 4
handLocations = [Location(300, 525), Location(75, 300), Location(300, 75), Location(525, 300), Location(300, 300)]
bidLocations = [Location(300, 350), Location(250, 300), Location(300, 250), Location(350, 300)]
stockLocations = [Location(400, 500), Location(100, 400), Location(200, 100), Location(500, 200)]

hands = None
bids = [0] * nbPlayers
stocks = [0] * nbPlayers
talon = None
currentPlayer = 0
nbMovesInRound = 0
targetCount = 0
  
cg = CardGame(600, 600, 30)
dealingOut()
cg.setStatusText("Dealing out...done. Double click to play.")
initBids()
initStocks()
hands[0].setTouchEnabled(True)

File: cardgame_tutorial/Ex06b.java
// Ex06b.java

import ch.aplu.jcardgame.*;
import ch.aplu.jgamegrid.*;
import ch.aplu.util.Monitor;

public class Ex06b extends CardGame
{
  public enum Suit
  {
    SPADES, HEARTS, DIAMONDS, CLUBS
  }

  public enum Rank
  {
    ACE, KING, QUEEN, JACK, TEN
  }

  class MyCardValues implements Deck.CardValues
  {
    public int[] values(Enum suit)
    {
      int[] defaultValues = new int[]
      {
        11, 4, 3, 2, 10
      };
      return defaultValues;
    }
  }
  //
  private final Deck deck =
    new Deck(Suit.values(), Rank.values(), "cover", new MyCardValues());
  private final int nbPlayers = 4;
  private final Location[] handLocations =
  {
    new Location(300, 525),
    new Location(75, 300),
    new Location(300, 75),
    new Location(525, 300),
    new Location(300, 300)
  };
  private final Location[] bidLocations =
  {
    new Location(300, 350),
    new Location(250, 300),
    new Location(300, 250),
    new Location(350, 300),
  };
  private final Location[] stockLocations =
  {
    new Location(400, 500),
    new Location(100, 400),
    new Location(200, 100),
    new Location(500, 200),
  };
  private Hand[] hands;
  private Hand[] bids = new Hand[nbPlayers];
  private Hand[] stocks = new Hand[nbPlayers];
  private Hand talon;
  private int currentPlayer = 0;
  private int nbMovesInRound = 0;
  private int targetCount = 0;

  public Ex06b()
  {
    super(600, 600, 30);
    dealingOut();
    setStatusText("Dealing out...done. Starting player: " + currentPlayer);
    initBids();
    initStocks();
    hands[0].setTouchEnabled(true);
  }

  private void dealingOut()
  {
    setStatusText("Dealing out...");

    hands = deck.dealingOut(nbPlayers, 0, false);  // All cards in talon
    talon = hands[nbPlayers];
    talon.setVerso(true);

    for (int i = 0; i < nbPlayers; i++)
    {
      hands[i].setView(this, new StackLayout(handLocations[i], 90 * i));
      hands[i].draw();
    }

    talon.setView(this, new StackLayout(handLocations[nbPlayers]));
    talon.draw();

    while (!talon.isEmpty())
    {
      for (int i = 0; i < nbPlayers; i++)
      {
        Card top = talon.getLast();
        top.setVerso(i == 0 ? false : true);
        talon.setTargetArea(new TargetArea(handLocations[i]));
        talon.transfer(top, hands[i], true);
      }
    }
  }

  private void initBids()
  {
    for (int i = 0; i < nbPlayers; i++)
    {
      bids[i] = new Hand(deck);
      bids[i].setView(this, new StackLayout(bidLocations[i]));
      bids[i].addCardListener(new CardAdapter()
      {
        public void atTarget(Card card, Location loc)
        {
          targetCount++;
          if (targetCount == nbPlayers)
            Monitor.wakeUp();  // All cards in stock->continue
        }
      });

      hands[i].setTargetArea(new TargetArea(bidLocations[i]));
      final int k = i;
      hands[i].addCardListener(new CardAdapter()
      {
        public void leftDoubleClicked(Card card)
        {
          card.setVerso(false);
          card.transfer(bids[k], true);
          hands[currentPlayer].setTouchEnabled(false);
          currentPlayer = (currentPlayer + 1) % nbPlayers;
          if (nbMovesInRound == 3)
          {
            setStatusText("Evaluating round...");
            nbMovesInRound = 0;
            currentPlayer = transferToWinner();
            stocks[currentPlayer].draw();
          }
          else
            nbMovesInRound++;
          if (hands[currentPlayer].isEmpty())
            showResult();
          else
          {
            setStatusText("Current player: " + currentPlayer);
            hands[currentPlayer].setTouchEnabled(true);
          }
        }
      });
    }
  }

  private void initStocks()
  {
    for (int i = 0; i < nbPlayers; i++)
    {
      stocks[i] = new Hand(deck);
      double rotationAngle;
      if (i == 0 || i == 2)
        rotationAngle = 0;
      else
        rotationAngle = 90;
      stocks[i].setView(this, new StackLayout(stockLocations[i], rotationAngle));
    }
  }

  private int transferToWinner()
  {
    delay(3000);
    Hand eval = new Hand(deck);
    for (int i = 0; i < nbPlayers; i++)
      eval.insert(bids[i].getFirst(), false);
    int nbWinner = eval.getMaxPosition(Hand.SortType.RANKPRIORITY);
    setStatusText("Round winner: " + nbWinner);
    transferToStock(nbWinner);
    return nbWinner;
  }

  private void transferToStock(int player)
  {
    targetCount = 0;
    for (int i = 0; i < nbPlayers; i++)
    {
      bids[i].setTargetArea(new TargetArea(stockLocations[player]));
      Card c = bids[i].getLast();
      if (c == null)
        continue;
      c.setVerso(true);
      bids[i].transferNonBlocking(c, stocks[player], true);
    }
    Monitor.putSleep();  // Wait until all cards are transferred to stock
    stocks[player].draw();
  }

  private void showResult()
  {
    String text = "Game over. Summary: ";
    for (int i = 0; i < nbPlayers; i++)
    {
      text += "Player # " + i + ": " + stocks[i].getScore();
      if (i < nbPlayers - 1)
        text += "; ";
    }
    setStatusText(text);
  }

  public static void main(String[] args)
  {
    new Ex06b();
  }
}
